src/providers/aws.rs

```diff
@@ -1227,53 +1254,53 @@ impl RegionLauncher {
     /// security groups often will not be deleted, due to timing quirks in the AWS api.
+    #[instrument(debug)]
     pub async fn shutdown(&mut self) {
         let client = self.client.as_ref().unwrap();
-        let log = self.log.as_ref().expect("RegionLauncher uninitialized");
         // terminate instances
         if !self.instances.is_empty() {
-            info!(log, "terminating instances");
+            tracing::info!("terminating instances");
             let instances = self.instances.keys().cloned().collect();
             self.instances.clear();
             let mut termination_req = rusoto_ec2::TerminateInstancesRequest::default();
             termination_req.instance_ids = instances;
             while let Err(e) = client.terminate_instances(termination_req.clone()).await {
-                let msg = format!("{}", e);
+                let msg = e.to_string();
                 if msg.contains("Pooled stream disconnected") || msg.contains("broken pipe") {
-                    trace!(log, "retrying instance termination");
+                    tracing::trace!("retrying instance termination");
                     continue;
                 } else {
-                    warn!(log, "failed to terminate tsunami instances: {:?}", e);
+                    tracing::warn!("failed to terminate tsunami instances: {}", e);
                     break;
                 }
             }
         }
 
-        debug!(log, "cleaning up temporary resources");
+        tracing::debug!("cleaning up temporary resources");
         if !self.security_group_id.trim().is_empty() {
-            trace!(log, "cleaning up temporary security group"; "name" => self.security_group_id.clone());
+            let group_span = tracing::trace_span!("security_group", id = %self.security_group_id);
+            let _guard = group_span.enter();
+
+            tracing::trace!("removing security group");
             // clean up security groups and keys
             // TODO need a retry loop for the security group. Currently, this fails
             // because AWS takes some time to allow the security group to be deleted.
             let mut req = rusoto_ec2::DeleteSecurityGroupRequest::default();
             req.group_id = Some(self.security_group_id.clone());
             if let Err(e) = client.delete_security_group(req).await {
-                warn!(log, "failed to clean up temporary security group";
-                    "group_id" => &self.security_group_id,
-                    "error" => ?e,
-                )
+                tracing::warn!("failed to clean up temporary security group: {}", e);
             }
         }
 
         if !self.ssh_key_name.trim().is_empty() {
-            trace!(log, "cleaning up temporary keypair"; "name" => self.ssh_key_name.clone());
+            let key_span = tracing::trace_span!("key", name = %self.ssh_key_name);
+            let _guard = key_span.enter();
```

Comment:
It looks like these span entered guards are held across await points, which is probably going to result in malformed traces. You should use `async` blocks and `.instrument` instead. See https://github.com/tokio-rs/tracing/#in-asynchronous-code

Reply:
Ah, that's really sad. I guess that means a lot more code indentation :'(

Reply:
That extra block also means an extra row in `eyre` backtraces I think, since there's now one more `?` (for the extra `async { ... }.await?`).

Reply:
Ended up being this: https://github.com/jonhoo/tsunami/pull/32/commits/084205c6991a6d86a473221cde5a7269da010eb1

Reply:
@yaahc tagging you specifically here for the observation about eyre backtraces with the extra `?` above.

Reply:
Going to leave this unresolved just so the discussion doesn't disappear entirely.